tests/test_vectorizer.py

```diff
@@ -0,0 +1,1044 @@
+              , _columns=>array['color', 'weight']
+              )
+            , 'public', 'thing'
+            )
+            """,
+            {
+                "implementation": "python_template",
+                "columns": ["color", "weight"],
+                "template": "color: $color weight: $weight $chunk",
+            },
+        ),
+    ]
+    bad = [
+        (
+            """
+            select ai._validate_formatting_python_template
+            ( ai.formatting_python_template
+              ( 'color: $color weight: $weight height: $height $chunk'
+              , _columns=>array['color', 'weight', 'height']
+              )
+            , 'public', 'thing'
+            )
+            """,
+            "columns in config do not exist in the table: height",
+        ),
+        (
+            """
+            select ai._validate_formatting_python_template
+            ( ai.formatting_python_template
+              ( 'color: $color weight: $weight height: $height' -- no $chunk
+              )
+            , 'public', 'thing'
+            )
+            """,
+            "template must contain $chunk placeholder",
+        ),
+        (
+            """
+            select ai._validate_formatting_python_template
+            ( ai.formatting_python_template
+              ( 'color: $color weight: $weight height: $height $chunk'
+              )
+            , 'public', 'thing2' -- has a column named "chunk"
+            )
+            """,
+            'formatting_python_template may not be used when source table has a column named "chunk"',
+        ),
+    ]
+    with psycopg.connect(db_url("test"), autocommit=True) as con:
+        with con.cursor() as cur:
+            cur.execute("drop table if exists public.thing;")
+            cur.execute("create table public.thing (id int, color text, weight float)")
+            cur.execute("drop table if exists public.thing2;")
+            cur.execute("create table public.thing2 (id int, color text, weight float, chunk text)")
+            for query, expected in ok:
+                cur.execute(query)
+                actual = cur.fetchone()[0]
+                assert actual.keys() == expected.keys()
+                for k, v in actual.items():
+                    assert k in expected and v == expected[k]
+            for query, err in bad:
+                try:
+                    cur.execute(query)
+                except psycopg.ProgrammingError as ex:
+                    msg = str(ex.args[0])
+                    assert len(msg) >= len(err) and msg[:len(err)] == err
+                else:
+                    pytest.fail(f"expected exception: {err}")
+
+
+VECTORIZER_ROW = r"""
+{
+    "id": 1,
+    "config": {
+        "chunking": {
+            "separator": "\n\n",
+            "chunk_size": 128,
+            "chunk_column": "body",
+            "chunk_overlap": 10,
+            "implementation": "character_text_splitter",
```

Comment:
can we separate this from the other attributes, otherwise validation will be complicated for me. 

maybe:

```json
"chunking" {
  "implementation": "character_text_splitter",
  "config": {
            "separator": "\n\n",
            "chunk_size": 128,
            "chunk_column": "body",
            "chunk_overlap": 10,
  }
}```

Same for embedding, and formatting.

Reply:
Using a discriminated union ought to work.